File: Hazarath/Classical_Beamforming/Classical_Beamforming.py
import numpy as np

# Generate a sinusoidal signal with AWG noise
def generate_signal(sampling_rate=10e6,
                    number_of_samples=5,
                    center_frequency=1e6):
    time = np.arange(number_of_samples) / sampling_rate

    # Create signal with AWG noise
    AWG_noise = np.random.normal(0, 0.1, number_of_samples) + 1j * np.random.normal(0, 0.1, number_of_samples)
    transmitter_signal = np.exp(2j * np.pi * center_frequency * time)
    transmitter_signal_with_noise = transmitter_signal + AWG_noise

    return time, AWG_noise, transmitter_signal, transmitter_signal_with_noise

# Delay-and-Sum Beamformer Implementation
def delay_and_sum_beamformer(time, AWG_noise, transmitter_signal, transmitter_signal_with_noise):
    "Construct the array steering vector"
    # Design the ULA with 3 element and half-wavelength spacing
    element_spacing = 0.5
    number_of_elements = 10

    # Define the direction of arrival of signals
    theta_degrees = 20
    # Convert the direction of arrival to radians
    theta = theta_degrees / 180 * np.pi

    # Compute the steering vector of the signal
    steering_vector = lambda theta, number_of_elements, element_spacing = element_spacing : np.exp(-2j * np.pi * element_spacing * np.arange(number_of_elements) * np.sin(theta))
    v_signal = steering_vector(theta, number_of_elements)[:, None]

    # To simulate the signal impinging on the array, we will matrix multiply the signal with the steering vector
    # and then add noise. In reality, each element will always be noisy and this is what we are simulating when
    # we add additional noise here which is denoted by σₑ.

    # X = s_tx * v_signal.T + σₑ
    # Where X is the impinging on the array
    AWG_noise_received = np.random.normal(loc=0, scale=0.1, size=(number_of_samples, number_of_elements)) + 1j * np.random.normal(loc=0, scale=0.1, size=(number_of_samples, number_of_elements))
    X = (transmitter_signal_with_noise[:, None] @ v_signal.T) + AWG_noise_received
    print(f"X.shape: {X.shape}")
    print(X)

    # Find the arrival angle
    # Define the angles to be searched
    thetas = np.arange(-90, 90 + 0.1, 0.05)

    # Define an output and response array to hold the results
    outputs = []
    responses = []

    # Iterate through each angle to find the angle of arrival
    for theta in thetas:
        theta *= np.pi/180

        # Steering vector
        w = steering_vector(theta, number_of_elements)

        # Process the signal at current angle
        y = (X @ w.conj()) / number_of_elements

        # Process the received signal with the steering vector
        # This processing is performed by matrix multiplying the Hermitian (Conjugate Transpose) of the
        # steering vector at angle v_s(theta) with the received signal matrix X, which produces an N dimensional vector y.

        # y = v_s(theta) * X

        # This matrix multiplication is where the time delays and signal summing occur, the steering vector actually
        # applies a time delay to the signal at each element. This is because there is a direct relation between time delay
        # and angle of arrival.

        # v_m(theta) = exp(-j * pi * m * sin(theta))

        # m * sin(theta) corresponds to the time delay based on physical location m of the element.
        # This time delay is proportional to angle theta

        # Compute array response
        # The array response is determined by the variance of the output vector y, if we have a small variance, then
        # we know that the received signals have a large degree of constructive interference and the output signal will be
        # the largest.

        # Compute power in dB
        array_response = 10 * np.log10(np.var(y))

        outputs.append(y)

        responses.append(array_response)

    # Normalize Array Responses
    responses -= np.max(responses)

    # Obtain angle that gives the maximum value
    angle_idx = np.argmax(responses)

    angle_of_arrival = thetas[angle_idx]

    print(f"Angle of arrival: {angle_of_arrival}")


if __name__ == "__main__":

    print("##### Introduction to Beamforming #####")
    # In this code, the Delay-and-Sum Beamformer is implemented from scratch

    # Definition:
    # Beamforming is a signal processing technique that aims to estimate the Direction of signals
    # impinging on a sensor array.

    # Beamforming is achieved by combining elements such that some add constructively and others add
    # destructively.

    # In this algorithm, I considered a Uniform Linear Array (ULA) with M number of elements.

    # If the element spacing is known, then the distance that the wavefront must travel at each subsequent
    # element can be calculated.

    # The DAS beamformer applies a time delay to the incoming signal from each element and sums the output together.
    # If we get the time delays correct, we will have a single high output signal.
    # We can then use the time delays that produced this signal to determine the angle of it's arrival.

    # In order to calculate the time delays, we can steer the array acoss multiple angles and choose the angle
    # that produces the largest response.

    # We process the incoming signal with the steering vector to produce an array response, the angle that produces
    # the largest response, which is most likely the ange of arrival.
    sampling_rate = 10e6
    number_of_samples = 5
    center_frequency = 1e6
    time, AWG_noise, transmitter_signal, transmitter_signal_with_noise = generate_signal(sampling_rate, number_of_samples, center_frequency)
    delay_and_sum_beamformer(time, AWG_noise, transmitter_signal, transmitter_signal_with_noise)

File: Hazarath/Classical_Beamforming/Classical_Beamforming.c
#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
#include <math.h>

void steering_vector(double Angle_rad, int Antenna_numbers,
                     double Antenna_dist, double complex *v)
{
    for (int i = 0; i < Antenna_numbers; i++)
    {
        v[i] = cexp(-2.0 * I * M_PI * Antenna_dist *
                    (double)i * sin(Angle_rad));
    }
}

double Delay_and_Sum_Beamforming(int num_samples, int Antenna_numbers,
                                 double Antenna_dist, double Min_Angle,
                                 double Max_Angle, double Angle_sum,
                                 double complex X[][10])
{
    int num_angles = (int)lround((Max_Angle - Min_Angle) / Angle_sum) + 1;

    double *angles = malloc(sizeof(double) * num_angles);
    double *responses = malloc(sizeof(double) * num_angles);

    double complex **outputs =
        malloc(sizeof(double complex *) * num_angles);

    for (int i = 0; i < num_angles; i++)
    {
        outputs[i] = malloc(sizeof(double complex) * num_samples);
    }

    /* Calculate the beam response for each scan angle. */
    for (int i = 0; i < num_angles; i++)
    {
        double Angle_deg = Min_Angle + (i * Angle_sum);
        angles[i] = Angle_deg;

        double Angle_rad = Angle_deg * (M_PI / 180.0);

        double complex w[Antenna_numbers];
        steering_vector(Angle_rad, Antenna_numbers, Antenna_dist, w);

        double complex *y = outputs[i];

        /* Apply the steering weights and sum the antenna signals. */
        for (int n = 0; n < num_samples; n++)
        {
            double complex s = 0.0;

            for (int m = 0; m < Antenna_numbers; m++)
            {
                s += X[n][m] * conj(w[m]);
            }

            y[n] = s / (double)Antenna_numbers;
        }

        /* Calculate the variance of the beamformed signal. */
        double complex mean_y = 0.0;

        for (int n = 0; n < num_samples; n++)
        {
            mean_y += y[n];
        }

        mean_y /= (double)num_samples;

        double var = 0.0;

        for (int n = 0; n < num_samples; n++)
        {
            double d = cabs(y[n] - mean_y);
            var += d * d;
        }

        var /= (double)num_samples;

        responses[i] = 10.0 * log10(var);
    }

    /* Find the angle with the maximum response. */
    double max_resp = responses[0];

    for (int i = 1; i < num_angles; i++)
    {
        if (responses[i] > max_resp)
        {
            max_resp = responses[i];
        }
    }

    for (int i = 0; i < num_angles; i++)
    {
        responses[i] -= max_resp;
    }

    int id = 0;

    for (int i = 1; i < num_angles; i++)
    {
        if (responses[i] > responses[id])
        {
            id = i;
        }
    }

    double angle_of_arrival = angles[id];

    printf("Angle of arrival: %.4f\n", angle_of_arrival);
}

int main()
{
    double complex X[5][10] = {
        {
             0.85389027 + 0.11896897 * I,
             0.69666460 - 0.71169760 * I,
            -0.29730123 - 0.90915896 * I,
            -0.87365642 + 0.07828388 * I,
            -0.58529659 + 1.11842390 * I,
             0.55546640 + 0.85601476 * I,
             1.02130987 - 0.09294449 * I,
             0.38525233 - 0.93593659 * I,
            -0.54866345 - 0.76254513 * I,
            -0.96715965 + 0.12439232 * I
        },

        {
             0.76664164 + 0.62969247 * I,
             0.85418111 - 0.41183408 * I,
            -0.05751092 - 1.04622193 * I,
            -0.66035444 - 0.58599790 * I,
            -0.98106196 + 0.49086581 * I,
            -0.05911355 + 0.69572811 * I,
             1.07131239 + 0.55763058 * I,
             0.82284905 - 0.65358731 * I,
            -0.22744990 - 0.77553594 * I,
            -0.96801588 - 0.36177181 * I
        },

        {
             0.22381305 + 1.04164448 * I,
             0.91015579 + 0.29067872 * I,
             0.67932165 - 0.74379019 * I,
            -0.46341275 - 0.93595163 * I,
            -1.25897993 - 0.11709991 * I,
            -0.61293528 + 0.72295844 * I,
             0.40359438 + 1.13886837 * I,
             1.05363762 + 0.09146186 * I,
             0.73039042 - 0.91903867 * I,
            -0.43858021 - 0.92023979 * I
        },

        {
            -0.29631709 + 1.15288117 * I,
             0.82430729 + 0.56756633 * I,
             0.84440535 - 0.24667412 * I,
             0.11560455 - 0.95365979 * I,
            -1.08070715 - 0.85376828 * I,
            -1.01858514 + 0.45139882 * I,
             0.04073744 + 1.17947305 * I,
             0.96934266 + 0.66940536 * I,
             1.00762776 - 0.51036050 * I,
            -0.00697367 - 1.04960891 * I
        },

        {
            -0.91355146 + 0.65446208 * I,
             0.23507138 + 1.22807686 * I,
             1.08702809 + 0.38506804 * I,
             0.94671503 - 0.89634679 * I,
            -0.49199050 - 1.23594469 * I,
            -1.16853178 - 0.40758970 * I,
            -0.89751429 + 0.87934276 * I,
             0.45095823 + 1.19469673 * I,
             1.29329044 + 0.23266057 * I,
             0.90291352 - 0.84050359 * I
        }
    };

    Delay_and_Sum_Beamforming(5, 10, 0.5, -90.0, 90.0, 0.05, X);

    return 0;
}
